Pullbacks of synthetic categories (#1183)
UniMath/agda-unimath: src/synthetic-category-theory/cospans-synthetic-categories.lagda.md

# Cospans of synthetic categories

```agda
{-# OPTIONS --guardedness #-}

module synthetic-category-theory.cospans-synthetic-categories where
```

<details><summary>Imports</summary>

```agda
open import foundation.cartesian-product-types
open import foundation.dependent-pair-types
open import foundation.universe-levels

open import structured-types.globular-types

open import synthetic-category-theory.synthetic-categories
```

</details>

## Idea

> A cospan is a pair of functors f, g with a common codomain:
> 
> ```text
> C --f--> E <--g-- D
> ```

Reviewer: This section should mention that we are working with functors of synthetic categories. For example:

````suggestion
A {{#concept "cospan" Disambiguation="Synthetic categories"}} of [synthetic categories](synthetic-category-theory.synthetic-categories.md) is a pair of functors f, g of synthetic categories with a common codomain:

```text
C --f--> E <--g-- D.
```
````

## Definition

```agda
module _
  {l : Level}
  where

  cospan-Synthetic-Category-Theory :
    (κ : language-Synthetic-Category-Theory l)
    (C E D : category-Synthetic-Category-Theory κ) → UU l
  cospan-Synthetic-Category-Theory κ C E D =
    Σ ( functor-Synthetic-Category-Theory κ C E)

>       λ f → functor-Synthetic-Category-Theory κ D E

Reviewer: Our convention is to put a pair of parentheses around the last lambda, even if the syntax of Agda technically allows us to omit them. We think it impairs the readability to omit them.

```suggestion
      ( λ f → functor-Synthetic-Category-Theory κ D E)
```

  right-source-cospan-Synthetic-Category-Theory κ {D = D} S = D

  target-cospan-Synthetic-Category-Theory :
    (κ : language-Synthetic-Category-Theory l)
    {C E D : category-Synthetic-Category-Theory κ} →
    cospan-Synthetic-Category-Theory κ C E D →
    category-Synthetic-Category-Theory κ
  target-cospan-Synthetic-Category-Theory κ {E = E} S = E

  left-map-cospan-Synthetic-Category-Theory :
    (κ : language-Synthetic-Category-Theory l)
    {C E D : category-Synthetic-Category-Theory κ}
    (S : cospan-Synthetic-Category-Theory κ C E D) →
    functor-Synthetic-Category-Theory κ
      ( left-source-cospan-Synthetic-Category-Theory κ S)
      ( target-cospan-Synthetic-Category-Theory κ S)
  left-map-cospan-Synthetic-Category-Theory κ = pr1

  right-map-cospan-Synthetic-Category-Theory :
    (κ : language-Synthetic-Category-Theory l)
    {C E D : category-Synthetic-Category-Theory κ}
    (S : cospan-Synthetic-Category-Theory κ C E D) →
    functor-Synthetic-Category-Theory κ
      ( right-source-cospan-Synthetic-Category-Theory κ S)
      ( target-cospan-Synthetic-Category-Theory κ S)
  right-map-cospan-Synthetic-Category-Theory κ = pr2
```

> ## Transformations of cone diagrams

Reviewer: ```suggestion
## Transformations of cospan diagrams
```

A transformation between cospans C --f--> E <--g-- D and C'--f'--> E' <--g'-- D'
is a commutative diagram of the form:

```text
C --f---> E <---g-- D
|         |         |
φ   τ⇙    χ    σ⇙   ψ
|         |         |
v         v         v
C'--f'--> E' <--g'--D'
```

```agda
module _
  {l : Level}
  where

>   transformation-cospan-Synthetic-Category-Theory :
>     (κ : language-Synthetic-Category-Theory l)
>     (μ : composition-Synthetic-Category-Theory κ)
>     {C C' E E' D D' : category-Synthetic-Category-Theory κ}
>     (S : cospan-Synthetic-Category-Theory κ C E D)
>     (S' : cospan-Synthetic-Category-Theory κ C' E' D') → UU l
>   transformation-cospan-Synthetic-Category-Theory κ μ S S' =
>     Σ ( functor-Synthetic-Category-Theory κ
>         ( left-source-cospan-Synthetic-Category-Theory κ S)
>         ( left-source-cospan-Synthetic-Category-Theory κ S'))
>     λ φ →
>       Σ ( functor-Synthetic-Category-Theory κ
>           ( target-cospan-Synthetic-Category-Theory κ S)
>           ( target-cospan-Synthetic-Category-Theory κ S'))
>       λ χ →
>         Σ ( functor-Synthetic-Category-Theory κ
>             ( right-source-cospan-Synthetic-Category-Theory κ S)
>             ( right-source-cospan-Synthetic-Category-Theory κ S'))
>         λ ψ →
>           Σ ( commuting-square-functors-Synthetic-Category-Theory κ μ
>               ( left-map-cospan-Synthetic-Category-Theory κ S)
>               ( χ)
>               ( φ)
>               ( left-map-cospan-Synthetic-Category-Theory κ S'))
>             λ τ →
>               commuting-square-functors-Synthetic-Category-Theory κ μ
>                 ( right-map-cospan-Synthetic-Category-Theory κ S)
>                 ( χ)
>                 ( ψ)
>                 ( right-map-cospan-Synthetic-Category-Theory κ S')

Reviewer: This was pretty good, but to make it exactly align with our coding conventions, it is better to indent the lambdas two spaces (indenting everything in its scope 2 further spaces too). 

Note that in the following formatting, the two arguments of each Sigma appear exactly below each other and nothing else. This makes it easier to spot how many arguments a function takes, and where those arguments are in the code.

```suggestion
  transformation-cospan-Synthetic-Category-Theory :
    (κ : language-Synthetic-Category-Theory l)
    (μ : composition-Synthetic-Category-Theory κ)
    {C C' E E' D D' : category-Synthetic-Category-Theory κ}
    (S : cospan-Synthetic-Category-Theory κ C E D)
    (S' : cospan-Synthetic-Category-Theory κ C' E' D') → UU l
  transformation-cospan-Synthetic-Category-Theory κ μ S S' =
    Σ ( functor-Synthetic-Category-Theory κ
        ( left-source-cospan-Synthetic-Category-Theory κ S)
        ( left-source-cospan-Synthetic-Category-Theory κ S'))
      ( λ φ →
        Σ ( functor-Synthetic-Category-Theory κ
            ( target-cospan-Synthetic-Category-Theory κ S)
            ( target-cospan-Synthetic-Category-Theory κ S'))
          ( λ χ →
            Σ ( functor-Synthetic-Category-Theory κ
                ( right-source-cospan-Synthetic-Category-Theory κ S)
                ( right-source-cospan-Synthetic-Category-Theory κ S'))
              ( λ ψ →
                Σ ( commuting-square-functors-Synthetic-Category-Theory κ μ
                    ( left-map-cospan-Synthetic-Category-Theory κ S)
                    ( χ)
                    ( φ)
                    ( left-map-cospan-Synthetic-Category-Theory κ S'))
                  ( λ τ →
                    commuting-square-functors-Synthetic-Category-Theory κ μ
                      ( right-map-cospan-Synthetic-Category-Theory κ S)
                      ( χ)
                      ( ψ)
                      ( right-map-cospan-Synthetic-Category-Theory κ S')))))
```
